File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

def dfs(s,current):
    global cnt
    next = G[s]
    if next in current:
        cnt+=1
        return 
    else:
        visited[next] = True
        current.append(next)
        dfs(next,current)



T = int(input())
for _ in range(T):
    cnt = 0
    n = int(input())
    G = list(map(int,input().split()))
    G = [x-1 for x in G]
    visited = [False for _ in range(n)]
    for i in range(n):
        if not visited[i]:
            dfs(i,[]) 
    print(cnt)

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java
import java.io.*;
import java.util.*;

public class Main {


    static int[] G;
    static boolean[] visited;
    static int cnt;

    public static void dfs(int s, List<Integer> current) {
        int next = G[s];
        if (current.contains(next)) {
            cnt++;
            return;
        } else {
            visited[next] = true;
            current.add(next);
            dfs(next, current);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();
        while (T-- > 0) {
            cnt = 0;
            int n = Integer.parseInt(br.readLine());
            G = new int[n];
            visited = new boolean[n];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                G[i] = Integer.parseInt(st.nextToken()) - 1;
            }

            for (int i = 0; i < n; i++) {
                if (!visited[i]) {
                    visited[i] = true;
                    List<Integer> current = new ArrayList<>();
                    current.add(i);
                    dfs(i, current);
                }
            }
            sb.append(cnt).append("\n");
        }
        System.out.println(sb);
    }
}



